apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -72,6 +81,60 @@ store_message(Msg) ->
 find_subscribers(_Msg) ->
     [node()].
 
+open_session(ClientID) ->
+    ?WHEN_ENABLED(emqx_ds:session_open(ClientID)).
+
+-spec add_subscription(emqx_types:topic(), emqx_ds:session_id()) ->
+    {ok, emqx_ds:iterator_id(), IsNew :: boolean()} | {skipped, disabled}.
+add_subscription(TopicFilterBin, DSSessionID) ->
+    ?WHEN_ENABLED(
+        begin
+            TopicFilter = emqx_topic:words(TopicFilterBin),
+            {ok, IteratorID, StartMS, IsNew} = emqx_ds:session_add_iterator(
+                DSSessionID, TopicFilter
+            ),
+            Ctx = #{
+                iterator_id => IteratorID,
+                start_time => StartMS,
+                is_new => IsNew
+            },
+            ?tp(persistent_session_ds_iterator_added, Ctx),
+            ?tp_span(
+                persistent_session_ds_open_iterators,
+                Ctx,
+                ok = open_iterator_on_all_nodes(TopicFilter, StartMS, IteratorID)
+            ),
+            {ok, IteratorID, IsNew}
+        end
+    ).
+
+-spec open_iterator_on_all_nodes(emqx_topic:words(), emqx_ds:time(), emqx_ds:iterator_id()) -> ok.
+open_iterator_on_all_nodes(TopicFilter, StartMS, IteratorID) ->
+    ?tp(persistent_session_ds_will_open_iterators, #{
+        iterator_id => IteratorID,
+        start_time => StartMS
+    }),
+    Nodes = emqx:running_nodes(),
```

Comment:
Maybe we could use something like emqx_cluster_rpc for this?

Reply:
I'm afraid no, because it's on the `emqx_conf`'s level. Besides, it's concerned with core nodes primarily, while we're working under different model where there's no such thing as core / replicant nodes.

Reply:
It sounds like this would need to be some kind of startup operation: during node startup, the node goes through all iterators from the session store (after waiting on that table) and opens them.  :thinking:

Reply:
Interesting, why? Sounds expensive, besides clients resuming sessions will open them eventually anyway. But I must be missing something.

Reply:
It's likely I'm misunderstanding something or being naive.

If we are to open the iterators on all nodes, even if _not running_, wouldn't we need something like this?

Reply:
Ah. No, we can't open the iterator on the node which is not running.

Yet, under our guarantees we can make progress (i.e. replay messages) if some of the nodes are temporarily unavailable, we'll replay from those still available. And then replay from those coming back online. What's important though is that we **must** consider a set of all cluster nodes, not just running at some specific point in time.

Reply:
Hmm I see.  I was thinking of that mechanism because of this:

> We probably want to open iterators on all nodes regardless of their status (i.e. _whether they are up or down_), at least eventually

So, indeed, those nodes coming back online will have to check with other nodes the existence of missed iterators, right?

Reply:
I currently think it's the session's job to see which nodes came back online and open / restore iterators there.